clients/src/main/java/org/apache/kafka/clients/consumer/PartitionAssignor.java

```diff
@@ -0,0 +1,125 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.kafka.clients.consumer;
+
+import org.apache.kafka.clients.consumer.internals.MetadataSnapshot;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.protocol.types.Type;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Generic interface for client-side assignment implementations. Unless custom metadata is
+ * needed, implementations will typically extend {@link org.apache.kafka.clients.consumer.internals.AbstractPartitionAssignor}.
+ * @param <M> The type of the metadata
+ */
+public interface PartitionAssignor<M> {
+
+    /**
+     * Perform the assignment for the consumer.
+     * @param consumerId The identifier of the local consumer instance
+     * @param consumers Map of all consumers in the group and their respective metadata
+     * @param metadataSnapshot The metadata snapshot that was used when the group was created, including
+     *                         subscription information and topic metadata
+     * @return
+     */
+    AssignmentResult assign(String consumerId,
+                            Map<String, M> consumers,
+                            MetadataSnapshot metadataSnapshot);
+
+    /**
+     * The name of the partition assignor (e.g. 'roundrobin')
+     * @return The name (must not be null)
+     */
+    String name();
+
+    /**
+     * The version of the partition assignor
+     * @return The version
+     */
+    short version();
+
+    /**
+     * Schema representing this assignor's metadata.
+     * @return Non-null schema
+     */
+    Type schema();
+
+    /**
+     * Get the local metadata for this assignor. This will be invoked prior to every
+     * round of group membership, so it is allowed to change dynamically. The metadata
+     * will probably include at least the subscription of the consumer, but can also include
+     * local information such as the number of processors on the host or its fqdn. Note,
+     * however, that extra care must be taken when designing an assignor for large groups
+     * since the metadata from each member must be propagated to ALL members of the group.
+     *
+     * @param metadataSnapshot Snapshot of the consumer's metadata/subscription.
+     * @return The latest metadata for
+     */
+    M metadata(MetadataSnapshot metadataSnapshot);
+
+    /**
+     * Wrapper
+     */
+    class AssignmentResult {
+        private boolean succeeded = false;
+        private List<TopicPartition> assignment;
+        private Set<String> groupSubscription;
+
+        public AssignmentResult(boolean succeeded,
```

Comment:
We could make this private since the `success` and `failure` methods seem to be the preferred way of constructing these objects and ensure you construct a valid combination (i.e. guarantees the right fields are null depending on the value of `succeeded`).